src/sources/audiosourcemp3.cpp

```diff
@@ -0,0 +1,445 @@
+                    mad_header_finish(&madHeader);
+                    return ERR; // abort
+                }
+            }
+        }
+
+        // Grab data from madHeader
+        const size_type madChannelCount = MAD_NCHANNELS(&madHeader);
+        if (kChannelCountDefault == getChannelCount()) {
+            // initially set the number of channels
+            setChannelCount(madChannelCount);
+        } else {
+            // check for consistent number of channels
+            if ((0 < madChannelCount) && getChannelCount() != madChannelCount) {
+                qWarning() << "Differing number of channels in some headers:"
+                        << m_file.fileName() << getChannelCount() << "<>"
+                        << madChannelCount;
+            }
+        }
+        const size_type madSampleRate = madHeader.samplerate;
+        if (kFrameRateDefault == getFrameRate()) {
+            // initially set the frame/sample rate
+            switch (madSampleRate) {
+            case 8000:
+                madUnits = MAD_UNITS_8000_HZ;
+                break;
+            case 11025:
+                madUnits = MAD_UNITS_11025_HZ;
+                break;
+            case 12000:
+                madUnits = MAD_UNITS_12000_HZ;
+                break;
+            case 16000:
+                madUnits = MAD_UNITS_16000_HZ;
+                break;
+            case 22050:
+                madUnits = MAD_UNITS_22050_HZ;
+                break;
+            case 24000:
+                madUnits = MAD_UNITS_24000_HZ;
+                break;
+            case 32000:
+                madUnits = MAD_UNITS_32000_HZ;
+                break;
+            case 44100:
+                madUnits = MAD_UNITS_44100_HZ;
+                break;
+            case 48000:
+                madUnits = MAD_UNITS_48000_HZ;
+                break;
+            default:
+                qWarning() << "Invalid sample rate:" << m_file.fileName()
+                        << madSampleRate;
+                return ERR; // abort
+            }
+            setFrameRate(madSampleRate);
+        } else {
+            // check for consistent frame/sample rate
+            if ((0 < madSampleRate) && (getFrameRate() != madSampleRate)) {
+                qWarning() << "Differing sample rate in some headers:"
+                        << m_file.fileName()
+                        << getFrameRate() << "<>" << madSampleRate;
+                return ERR; // abort
+            }
+        }
+
+        mad_timer_add(&madDuration, madHeader.duration);
+        sumBitrate += madHeader.bitrate;
+        mad_header_finish(&madHeader);
+
+        // Add next frame to list of frames
+        addSeekFrame(madDuration, madUnits, m_madStream.this_frame);
+    }
+
+    if (1 < m_seekFrameList.size()) {
+        setFrameCount(mad_timer_count(madDuration, madUnits));
+        m_avgSeekFrameCount = getFrameCount() / (m_seekFrameList.size() - 1);
```

Comment:
m_avgSeekFrameCount is now only used as a starting point for the binary search. It doesn't matter how accurate it is, the binary search will reliably find the frame we are looking for in O(log n). The binary search should be fast enough even without this initial guess.